orttraining/orttraining/test/python/checkpoint/orttraining_test_load_checkpoint.py

```diff
@@ -0,0 +1,1384 @@
+        # fp32 states are not sharded
+        for key, value in state_dict_post_checkpoint['fp16_param'].items():
+            assert_allclose(value, state_dict_pre_checkpoint['fp32_param'][key[:-5]], atol=1e-3)
+
+        # collect optimizer states for later comparison since they are sharded
+        aggregate_states(optimizer_states, state_dict_pre_checkpoint['optimizer'])
+
+    # compare optimizer states
+    for key, value in optimizer_states.items():
+        assert_allclose(value.reshape(state_dict_post_checkpoint['optimizer'][key].size()), state_dict_post_checkpoint['optimizer'][key])
+
+@distributed_setup
+def test_load_from_single_node_full_precision_into_distributed_zero_full_precision(world_rank, world_size, device, checkpoint_dir = 'checkpoint_dir/single_node/full_precision/'):
+    opts = {
+                'device' : {'id' : device},
+                'distributed' :
+                {
+                    'world_rank' : world_rank,
+                    'world_size' : world_size,
+                    'allreduce_post_accumulation' : True,
+                    'deepspeed_zero_optimization':
+                    {
+                        'stage': 1
+                    }
+                },
+                'debug' : {'deterministic_compute': True}
+            }
+    
+    # extract state dictionaries to compare
+    state_dict_post_checkpoint, model = load_evaluate_extract_state_dict(device, orttrainer.ORTTrainerOptions(opts), checkpoint_dir)
+    state_dict_post_checkpoint = split_state_dict(state_dict_post_checkpoint)
+
+    state = pickle.load(open(os.path.join(checkpoint_dir, 'state_dict.pkl'), 'rb'))
+    state_dict_pre_checkpoint = split_state_dict(state['state_dict'])
+
+    # compare all states
+    # model states
+    for key, value in state_dict_pre_checkpoint['fp32_param'].items():
+        assert_allclose(value, state_dict_post_checkpoint['fp32_param'][key])
+
+    # round about way of checking optimizer states. Save state dicts into temporary folder, read them and aggregate them.
+    pickle.dump(state_dict_post_checkpoint, open(os.path.join(checkpoint_dir, 'distributed_state_'+str(world_rank)+'.pkl'), "wb"))
+    dist.barrier()
+
+    if world_rank == 0:
+        num_states = len(glob.glob1(checkpoint_dir,"distributed_state*"))
+        optimizer_states = dict()
+        for rank in range(num_states):
+            rank_state_dict = pickle.load(open(os.path.join(checkpoint_dir, 'distributed_state_'+str(rank)+'.pkl'), 'rb'))
+
+            # collect optimizer states for later comparison since they are sharded
+            aggregate_states(optimizer_states, rank_state_dict['optimizer'])
+
+        # compare optimizer states
+        """
+        failure_case: non sharded to sharded weights fails
+        for key, value in optimizer_states.items():
+            assert_allclose(value.reshape(state_dict_pre_checkpoint['optimizer'][key].size()), state_dict_pre_checkpoint['optimizer'][key])
+        """
```

Comment:
What does this mean? This test would fail because of a bug?

Reply:
Ideally we dont let codes commented out as they get stale and forgotten. But as this will be short-lived, you can convert it into a TODO (like  TODO: Uncomment this after Checkpoint redesign. Current implementation does not support it) and let it commented out

Reply:
Added a TODO there.